src/journal/journald-stream.c

```diff
@@ -610,7 +610,17 @@ static int stdout_stream_new(sd_event_source *es, int listen_fd, uint32_t revent
         }
 
         if (s->n_stdout_streams >= STDOUT_STREAMS_MAX) {
-                log_warning("Too many stdout streams, refusing connection.");
+                struct ucred u;
+
+                r = getpeercred(fd, &u);
+
+                /* By closing fd here we make sure that the client won't wait too long for journald to
+                 * gather all the data it adds to the error message to find out that the connection has
+                 * just been refused.
+                 */
+                fd = safe_close(fd);
+
+                server_driver_message(s, r < 0 ? 0 : u.pid, NULL, LOG_MESSAGE("Too many stdout streams, refusing connection."), NULL);
```

Comment:
Couldn't you just use `u.pid` without the ternary dance if you initializ `u` to `{}`?  Then you could also `(void)` the return from `getpeercred()` instead of bothering with assigning it to `r`...

Reply:
@vcaputo thank you for the review! I agree that it'd be shorter, but it'd also be a little bit more fragile in the sense that it'd be easy to break it by accidentally removing the initialization. Given that there're a lot of fans of semantic patches applied automatically with Cocci who are [unstoppable](https://github.com/systemd/systemd/pull/8210) :-), I'd leave it as is but if everybody thinks that the alternative approach is better I can change it of course.

Reply:
Having written the following snippet, I think it's a bit harder to understand due to the implicit magic happening behind the scenes, to be honest:
```c
struct ucred u = {};
(void) getpeercred(fd, &u);
...
server_driver_message(s, u.pid, ...);
```
It's not clear why the errors are ignored, why `u.pid` is passed to `server_driver_message` anyway. `r < 0 ? 0 : u.pid` is straightforward, raises no questions and, more importantly, explicitly shows that `0` is passed to `server_driver_message` intentionally because it's special.

Reply:
@vcaputo by the way, in https://github.com/systemd/systemd/pull/10444#discussion_r226335993 @poettering and I briefly discussed a race happening when the client successfully writes some data to its end of the connection before `journald` closes its end, which means that the only way to figure out that the connection has been refused is to get `EPIPE` and `SIGPIPE` (which isn't the best API if you ask me :-)). I've been thinking how to synchronize clients and `journald` in a more predictable way, but, unfortunately, haven't come up with anything useful yet. Could you take a fresh look there? It's possible that I'm missing something obvious.

Reply:
It's a matter of taste, though perhaps your point about Cocci trumps all.

I personally find the shorter version clearer and concise,  since I don't see empty initializers as implicit magic.  It reads as explicitly initializing everything in `u` to zero, immediately followed by something which potentially modifies `u` - clearly optional thanks to the (void) cast on the return.  There's no ambiguity there in my mind; `u.pid` is zero or a valid pid, it all looks deliberate.

I'll take a look at #10444

Reply:
Thank you!

Indeed, at the end of the day, It's a matter of taste. Anyway, I prepared https://github.com/evverx/systemd/commit/452cbe377812fc0fd4a2944b84e59d7307e589cd, which I'm ready to force-push here at any time if necessary. I guess we just need someone else to review the PR to have a quorum :-)